File: algorithms/[785]判断二分图/solution.py
# topics = ["广度优先搜索"]

from collections import deque
from typing import Deque, List


class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        q: Deque[int] = deque()
        # 记录每个顶点所在子集（1 或者 2）, 初始时为 0
        id = [0] * n

        while q or id.count(0):
            if not q:
                i = id.index(0)
                id[i] = 1
                q.append(i)

            node = q.popleft()
            next = 2 if id[node] == 1 else 1

            for neighbor in graph[node]:
                if not id[neighbor]:
                    id[neighbor] = next
                    q.append(neighbor)
                elif id[neighbor] != next:
                    return False

        return True

File: algorithms/[785]判断二分图/solution.ts
// topics = ["深度优先搜索"]

function isBipartite(graph: number[][]): boolean {
  const n = graph.length;
  // 记录每个顶点所在子集（1 或者 2）, 初始时为 0
  const id = new Array<number>(n).fill(0);

  let res = true;

  const dfs = (node: number, s = 1): void => {
    id[node] = s;
    const next = s == 1 ? 2 : 1;
    for (const neighbor of graph[node]) {
      if (id[neighbor] == 0) {
        dfs(neighbor, next);
        if (!res) {
          return;
        }
      } else if (id[neighbor] != next) {
        res = false;
        return;
      }
    }
  };

  for (let i = 0; i < n; i += 1) {
    if (id[i] == 0) {
      dfs(i);
      if (!res) {
        break;
      }
    }
  }

  return res;
}

function isBipartite2(graph: number[][]): boolean {
  const n = graph.length;
  // 记录每个顶点所在子集（1 或者 2）, 初始时为 0
  const id = new Array<number>(n).fill(0);

  const dfs = (node: number, s = 1): boolean => {
    id[node] = s;
    const next = s == 1 ? 2 : 1;
    for (const neighbor of graph[node]) {
      if (id[neighbor] == 0 && !dfs(neighbor, next)) {
        return false;
      } else if (id[neighbor] != next) {
        return false;
      }
    }

    return true;
  };

  for (let i = 0; i < n; i += 1) {
    if (id[i] == 0 && !dfs(i)) {
      return false;
    }
  }

  return true;
}
